src/targets/gpu/mlir_conv.cpp

```diff
@@ -0,0 +1,264 @@
+    {
+        execution_spec *result = nullptr;
+
+#ifdef    MIGRAPHX_MLIR_MIOPEN_SUPPORT
+        auto conv  = any_cast<op::convolution>(op_r->get_operator());
+        auto inp_t = op_r->inputs().at(0)->get_shape();
+        auto flt_t = op_r->inputs().at(1)->get_shape();
+        auto out_t = op_r->get_shape();
+
+        auto get_type_str = [](const shape &s) -> const char * {
+          switch (s.type()) {
+          case shape::float_type: return "f32";
+          case shape::half_type: return "f16";
+          default: break;
+          }
+          return nullptr;
+        };
+
+        auto *inp_t_s = get_type_str(inp_t);
+        auto *flt_t_s = get_type_str(flt_t);
+        auto *out_t_s = get_type_str(out_t);
+
+        if (out_t_s == nullptr || inp_t_s == nullptr || flt_t_s == nullptr)
+          return nullptr;
+
+        std::string mlir_options =
+          " --kernel_name " + std::string(mlir_kernel_name);
+
+        // platform spec
+        auto &device = get_context().get_current_device();
+        mlir_options += 
+          " --arch " + device.get_device_name() +
+          " --num_cu " + std::to_string(device.get_cu_count()); // ???
+
+        // Conv spec
+        mlir_options += 
+          " --operation "   "conv2d"
+          " --batchsize " + std::to_string(conv.group) +
+          " --groupsize " + std::to_string(1) +
+          " --padding_h " + std::to_string(conv.padding[0]) +
+          " --padding_w " + std::to_string(conv.padding[1]) +
+          " --conv_stride_h " + std::to_string(conv.stride[0]) +
+          " --conv_stride_w " + std::to_string(conv.stride[1]) +
+          " --dilation_h " + std::to_string(conv.dilation[0]) +
+          " --dilation_w " + std::to_string(conv.dilation[1]);
+
+        // Input spec
+        mlir_options += 
+          " --in_layout "     "NGCHW"
+          " --in_type "     + std::string(inp_t_s) +
+          " --in_channels " + std::to_string(inp_t.lens()[1]) +
+          " --in_h "        + std::to_string(inp_t.lens()[2]) +
+          " --in_w "        + std::to_string(inp_t.lens()[3]);
+        
+        // Filter spec
+        mlir_options += 
+          " --fil_layout "    "NGCHW"
+          " --fil_type "    + std::string(flt_t_s) +
+          " --fil_h "       + std::to_string(flt_t.lens()[2]) +
+          " --fil_w "       + std::to_string(flt_t.lens()[3]);
+        
+        // Output spec
+        mlir_options += 
+          " --out_layout "     "NGCHW"
+          " --out_type "     + std::string(out_t_s) +
+          " --out_channels " + std::to_string(out_t.lens()[1]) +
+          " --out_h "        + std::to_string(out_t.lens()[2]) +
+          " --out_w "        + std::to_string(out_t.lens()[3]);
+
+        auto bin_i = binary_map.find(mlir_options);
+        if (bin_i == binary_map.end()) {
+          size_t bin_size = 0;
+          char *bin_buffer = nullptr;
+
+          auto mlir_handle = miirCreateHandle(mlir_options.c_str());
```

Comment:
The handle should be created with a managed pointer(which is just a `std::unique_ptr` with a custom deleter) so it will cleanup even if an exception occurs:

```cpp
using mlir_handle = MIGRAPHX_MANAGE_PTR(MiirHandle, miirDestroyHandle);

auto handle = mlir_handle{mlir_options.c_str()};
```

Reply:
How does that call miirCreateHandle?

Reply:
Oops, that should be `auto handle = mlir_handle{miirCreateHandle(mlir_options.c_str())};`